shared/src/ledger/mod.rs

```diff
@@ -14,6 +14,531 @@ pub mod queries;
+            tx: self.tx_builder(),
+        }
+    }
+
+    /// Make a Withdraw builder from the given minimum set of arguments
+    fn new_add_erc20_transfer(
+        &self,
+        sender: Address,
+        recipient: EthAddress,
+        asset: EthAddress,
+        amount: InputAmount,
+    ) -> args::EthereumBridgePool {
+        args::EthereumBridgePool {
+            sender,
+            recipient,
+            asset,
+            amount,
+            fee_amount: InputAmount::Unvalidated(token::DenominatedAmount {
+                amount: token::Amount::default(),
+                denom: NATIVE_MAX_DECIMAL_PLACES.into(),
+            }),
+            fee_payer: None,
+            fee_token: self.native_token(),
+            nut: false,
+            code_path: PathBuf::from(TX_BRIDGE_POOL_WASM),
+            tx: self.tx_builder(),
+        }
+    }
+
+    /// Make a ResignSteward builder from the given minimum set of arguments
+    fn new_resign_steward(&self, steward: Address) -> args::ResignSteward {
+        args::ResignSteward {
+            steward,
+            tx: self.tx_builder(),
+            tx_code_path: PathBuf::from(TX_RESIGN_STEWARD),
+        }
+    }
+
+    /// Make a UpdateStewardCommission builder from the given minimum set of
+    /// arguments
+    fn new_update_steward_rewards(
+        &self,
+        steward: Address,
+        commission: Vec<u8>,
+    ) -> args::UpdateStewardCommission {
+        args::UpdateStewardCommission {
+            steward,
+            commission,
+            tx: self.tx_builder(),
+            tx_code_path: PathBuf::from(TX_UPDATE_STEWARD_COMMISSION),
+        }
+    }
+
+    /// Make a TxCustom builder from the given minimum set of arguments
+    fn new_custom(&self, owner: Address) -> args::TxCustom {
+        args::TxCustom {
+            owner,
+            tx: self.tx_builder(),
+            code_path: None,
+            data_path: None,
+            serialized_tx: None,
+        }
+    }
+
+    /// Sign the given transaction using the given signing data
+    async fn sign(
+        &self,
+        tx: &mut Tx,
+        args: &args::Tx,
+        signing_data: SigningTxData,
+    ) -> crate::sdk::error::Result<()> {
+        signing::sign_tx(*self.wallet_mut().await, args, tx, signing_data)
+    }
+
+    /// Process the given transaction using the given flags
+    async fn submit(
+        &self,
+        tx: Tx,
+        args: &args::Tx,
+    ) -> crate::sdk::error::Result<ProcessTxResponse> {
+        tx::process_tx(self, args, tx).await
+    }
+}
+
+/// Provides convenience methods for common Namada interactions
+pub struct NamadaImpl<'a, C, U, V, I>
+where
+    C: crate::ledger::queries::Client + Sync,
+    U: WalletIo,
+    V: ShieldedUtils,
+    I: Io,
+{
+    /// Used to send and receive messages from the ledger
+    pub client: &'a C,
+    /// Stores the addresses and keys required for ledger interactions
+    pub wallet: RwLock<&'a mut Wallet<U>>,
```

Comment:
Unfortunately, I don't think the lock on the wallet makes it parallelizable as one task might overwrite the changes of another task because in-memory changes are not synced to fs without an explicit `save()` call. We have an issue to fix this (#50 - the wallet should lock internally during any mutable operations on it and only release the lock after implicitly saving to fs with as small scope as possible to avoid contention), but I'd like to avoid the illusion of this being thread-safe before that's done as it can lead to loss of keys. I'm not quite sure what to do with this until then

Reply:
Agreed, this PR buys us asynchrony within a Namada application at the cost of safety for multiple running Namada applications. (We might use it in this parallel application for instance: https://github.com/anoma/namada-sdk-starter/blob/murisi/updating-sdk-usage/src/main.rs .) Since this PR does fix SDK and API issues and the locking issue is not so trivial, could we perhaps merge this PR and then follow up with a fix for wallet safety?

One quick way to address the key loss issue may be to leave the API as-is and modify the `save` method of the wallet and shielded context to: 1) obtain lock on file 2) load the file into memory 3) merge the new changes into that file 4) save the merged file 5) release the lock. See https://github.com/anoma/namada/blob/3f979bf3b2707db4d065a4fae6e8c2340ae1d54b/shared/src/sdk/masp.rs#L633 . Then we will have more time to implement https://github.com/anoma/namada/issues/50 while knowing that keys are not being lost.

Reply:
That's a great idea! I think we should go with that